stdlib/mutex.mli

```diff
@@ -58,3 +58,10 @@ val unlock : t -> unit
 
    @before 4.12 {!Sys_error} was not raised when unlocking an unlocked mutex
    or when unlocking a mutex from a different thread. *)
+
+val with_locked : t -> (unit -> 'a) -> 'a
+(** [with_locked mutex f] runs [f()] in a critical section where [mutex]
+    is locked (using {!lock}); it then takes care of releasing [mutex],
+    whether [f()] returned a value or raised an exception.
```

Comment:
Can we have some guarantee in the interface that this function is resource-safe against asynchronous exceptions? (there is already `Fun.protect` which explicitly mentions that it does _not_ protect against asynchronous exceptions).

Reply:
I think I need an explanation of what async exceptions are, at this point. I've written OCaml for a while but this is not clear at all to me.

Reply:
I guess https://discuss.ocaml.org/t/ann-memprof-limits-preview-and-a-guide-to-handle-asynchronous-exceptions/5756 is a good starting point?

Reply:
The doc of `Fun.protect` gives the example of `exceptions raised from signal handlers (e.g. {!Sys.Break})`, I think this is good enough for a description in the documentation. The memprof-limits guide talks instead about exceptions raised from Memprof callbacks. As to how you avoid them, you simply make sure that there can be no polling location for asynchronous callbacks (allocation points, `caml_enter_blocking_section`, and the new polling locations introduced at the start of functions in other places; plus a few more in the case of bytecode).